components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultOIDCClaimsCallbackHandler.java

```diff
@@ -0,0 +1,628 @@
+     * @return
+     */
+    private static String getServiceProviderMappedUserRoles(ServiceProvider serviceProvider,
+                                                            List<String> locallyMappedUserRoles,
+                                                            String claimSeparator) throws FrameworkException {
+
+        if (CollectionUtils.isNotEmpty(locallyMappedUserRoles)) {
+            // Get Local Role to Service Provider Role mappings
+            RoleMapping[] localToSpRoleMapping = serviceProvider.getPermissionAndRoleConfig().getRoleMappings();
+
+            if (ArrayUtils.isNotEmpty(localToSpRoleMapping)) {
+                for (RoleMapping roleMapping : localToSpRoleMapping) {
+                    // check whether a local role is mapped to service provider role
+                    if (locallyMappedUserRoles.contains(roleMapping.getLocalRole().getLocalRoleName())) {
+                        // remove the local role from the list of user roles
+                        locallyMappedUserRoles.remove(roleMapping.getLocalRole().getLocalRoleName());
+                        // add the service provider mapped role
+                        locallyMappedUserRoles.add(roleMapping.getRemoteRole());
+                    }
+                }
+            }
+        }
+
+        return StringUtils.join(locallyMappedUserRoles, claimSeparator);
+    }
+
+    private Map<String, Object> getClaimsFromUserStore(OAuthAuthzReqMessageContext requestMsgCtx)
+            throws IdentityApplicationManagementException, IdentityException, UserStoreException {
+
+        Map<String, Object> mappedAppClaims = new HashMap<>();
+
+        String spTenantDomain = (String) requestMsgCtx.getProperty(MultitenantConstants.TENANT_DOMAIN);
+
+        // There are certain flows where tenant domain is not added as a message context property.
+        if (spTenantDomain == null) {
+            spTenantDomain = requestMsgCtx.getAuthorizationReqDTO().getTenantDomain();
+        }
+
+        ApplicationManagementService applicationMgtService = OAuth2ServiceComponentHolder.getApplicationMgtService();
+        String spName = applicationMgtService
+                .getServiceProviderNameByClientId(requestMsgCtx.getAuthorizationReqDTO().getConsumerKey(),
+                        INBOUND_AUTH2_TYPE, spTenantDomain);
+        ServiceProvider serviceProvider = applicationMgtService.getApplicationExcludingFileBasedSPs(spName,
+                spTenantDomain);
+        if (serviceProvider == null) {
+            return mappedAppClaims;
+        }
+        ClaimMapping[] requestedLocalClaimMap = serviceProvider.getClaimConfig().getClaimMappings();
+        if (requestedLocalClaimMap == null || !(requestedLocalClaimMap.length > 0)) {
+            return new HashMap<>();
+        }
+
+        AuthenticatedUser user = requestMsgCtx.getAuthorizationReqDTO().getUser();
+        String userTenantDomain = user.getTenantDomain();
+        String username = user.toString();
```

Comment:
Better to construct the username from AuthenticatatedUser rather than depend on toString() method.